src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,211 @@
+   into the executor.
+1. A kernel callback may call `Waker::wake` after its future finishes executing,
+   and the executor must ignore those wakeups. This duplicates the "ignore
+   spurious wakeups" functionality that exists in the future itself.
+
+Ultimately, this made the executor logic [quite
+nontrivial](https://github.com/tock/design-explorations/blob/master/size_comparison/futures/src/task.rs).
+The executor logic can be monomorphized for each future, which allows the
+compiler to make inlining optimizations, but results in a significant amount of
+duplicate code. Alternatively, it could be adapted to use function pointers or
+vtables to avoid the code duplication, but then the compiler *definitely* cannot
+inline `Future::poll` into the kernel callbacks.
+
+At this point, Barbara realizes she needs to do some benchmarking. Barbara comes
+up with a sample application -- an app that blinks a led and responds to button
+presses -- and implements it twice. One implementation does not use `Future` at
+all, the other does. In addition to two async drivers (a timer driver and a GPIO
+driver), the `Future`-based app uses Barbara's executor as well as a single
+combinator (that manages the state of the app).
+
+Barbara publishes an
+[analysis](https://github.com/tock/design-explorations/tree/master/size_comparison)
+of the relative sizes of the two app implementations, finding a large percentage
+increase in both code size and RAM usage (note: stack usage was not
+investigated). Barbara concludes that **`core::future::Future` is not suitable
+for highly-resource-constrained environments** due to the amount of code and RAM
+required to implement executors and combinators.
+
+Barbara redesigns the library she is building to use a [different
+concept](https://github.com/tock/design-explorations/tree/master/zst_pointer_async)
+for implementing async APIs in Rust that are much lighter weight. She has moved
+on from `Future` and is refining her [async
+traits](https://github.com/tock/libtock-rs/blob/master/core/platform/src/async_traits.rs)
+instead.
+
+## 🤔 Frequently Asked Questions
+
+* **What are the morals of the story?**
+    * `core::future::Future` isn't suitable for every asynchronous API in Rust.
+      `Future` has a lot of capabilities, such as the ability to spawn
+      dynamically-allocated futures, that are unnecessary in embedded system.
+    * We should look at embedded Rust's relationship with `Future` so we don't
+      fragment the embedded Rust ecosystem. Other embedded crates use `Future`,
+      presumably because they are less space constrained than Barbara's
+      codebase.
+* **Why did you choose *Barbara* to tell this story?**
+    * This story is about someone who is an experienced systems programmer and
+      an experienced Rust developer. All the other characters have "new to Rust"
+      or "new to programming" as a key characteristic.
+    * [Grace] could as worked as well.
```

Comment:
Grace appears twice in this list, I believe this first is the superfluous one.

Reply:
There is a weird juxtaposition between the fact that I feel Grace would have worked for the story as well and the fact I said the story may play out slightly differently with a Grace. Is that what you are referring to when you say that Grace appears twice?

I'm not quite sure how to fix it. It may actually be better to remove the statement that it would have played out differently with Grace, as I'm not sure that Grace's inexperience with Rust would have made a difference to how things played out.

Reply:
Ah, yes that was my confusion. To my reading, this first one is not necessary (sure, you _could_ have used Grace, but the explanation for why it would have gone differently sufficiently explains why Grace would have been a poorer choice). But idk :)

Reply:
I thought about it for another while, and removed the statement that Grace would have sufficed as well. I'm less confident in that statement than I am in my statement on how it may have played out differently with Grace.